source/extensions/quic/crypto_stream/envoy_tls_server_handshaker.cc

```diff
@@ -0,0 +1,75 @@
+#include "envoy_tls_server_handshaker.h"
+
+namespace Envoy {
+namespace Quic {
+
+EnvoyTlsServerHandshaker::EnvoyTlsServerHandshaker(
+    quic::QuicSession* session, const quic::QuicCryptoServerConfig* crypto_config,
+    OptRef<const Network::DownstreamTransportSocketFactory> transport_socket_factory,
+    Envoy::Event::Dispatcher& dispatcher)
+    : quic::TlsServerHandshaker(session, crypto_config) {
+  if (transport_socket_factory != absl::nullopt) {
+    transport_socket_factory_.emplace(
+        dynamic_cast<const QuicServerTransportSocketFactory&>(transport_socket_factory.ref()));
+    for (auto cert_config : transport_socket_factory_->getTlsCertificates()) {
+      if (cert_config.get().privateKeyMethod()) {
+        cert_config.get().privateKeyMethod()->registerPrivateKeyMethod(GetSsl(), *this, dispatcher);
+      }
+    }
+  }
+}
+
+ssl_private_key_result_t EnvoyTlsServerHandshaker::PrivateKeySign(uint8_t* out, size_t* out_len,
+                                                                  size_t max_out, uint16_t sig_alg,
+                                                                  absl::string_view in) {
+  if (transport_socket_factory_->getTlsCertificates().size() > 0) {
+    // TODO(soulxu): Currently the QUIC transport socket only support one certificate. After
+    // QUIC transport socket with multiple certificates, we will figure out how to support
+    // multiple certificates for private key provider also.
+    auto private_key_method =
+        transport_socket_factory_->getTlsCertificates()[0].get().privateKeyMethod();
```

Comment:
@ggreenway, It seems like I only need to invoke `getTlsCertificateAndKey` method of transport socket factory. And then I only to figure out how to get the sni hostname. I should be able get that from:
https://github.com/google/quiche/blob/02047e04d076d36226e92847a94acf33544f10d6/quiche/quic/core/tls_server_handshaker.h#L100-L101